File: python/src/anysearch/providers/serpapi.py
"""SerpApi — real-time Google (and other engines) SERP scraping."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability


class SerpApiProvider(BaseProvider):
    name = "serpapi"
    aliases = ("serp",)
    env_keys = ("SERPAPI_API_KEY", "SERPAPI_KEY")
    default_base_url = "https://serpapi.com"
    extra_package = "google-search-results"
    native_client = ("serpapi", ("Client", "GoogleSearch"))
    capabilities = frozenset(
        {
            Capability.COUNTRY,
            Capability.LANGUAGE,
            Capability.SAFE_SEARCH,
            Capability.NEWS,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        params = {
            "engine": "google",
            "q": req.query,
            "num": req.max_results,
            "api_key": self.api_key,
        }
        if req.country:
            params["gl"] = req.country.lower()
        if req.language:
            params["hl"] = req.language
        if req.safe_search:
            params["safe"] = "off" if req.safe_search == "off" else "active"
        if req.search_type == "news":
            params["tbm"] = "nws"
        params.update(req.extra)
        return PreparedRequest(
            method="GET",
            url=f"{self.base_url}/search",
            headers={"Accept": "application/json"},
            params=params,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        if req.search_type == "news":
            items = data.get("news_results", []) or []
        else:
            items = data.get("organic_results", []) or []
        results = []
        for item in items:
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("link"),
                    snippet=item.get("snippet"),
                    source=item.get("source"),
                    published_date=item.get("date"),
                    score=item.get("position"),
                    raw=item,
                )
            )
        answer = None
        box = data.get("answer_box") or {}
        if isinstance(box, dict):
            answer = box.get("answer") or box.get("snippet")
        total = (data.get("search_information") or {}).get("total_results")
        return self._response(
            req, results, data, answer=answer, total_results=total, elapsed_ms=elapsed_ms
        )

File: js/src/providers/serpapi.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";
import {
  applySerpapiLocale,
  applySerpapiSafeSearch,
  extractSerpapiAnswer,
  extractSerpapiResults,
  normalizeEngine,
  resolveSerpapiEngine,
  SERPAPI_WEB_ENGINES,
} from "./serpapiEngines.js";

export const serpapi: ProviderSpec = {
  name: "serpapi",
  aliases: ["serp"],
  envKeys: ["SERPAPI_API_KEY", "SERPAPI_KEY"],
  defaultBaseUrl: "https://serpapi.com",
  extraPackage: "serpapi",
  nativeImport: { module: "serpapi", exports: ["getJson", "GoogleSearch"] },
  capabilities: new Set(["country", "language", "safe_search", "news", "engine"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const defaultEngine = normalizeEngine(
      String(ctx.config?.engine ?? ctx.config?.defaultEngine ?? "google"),
    );
    const extra = { ...req.extra };
    const engine = resolveSerpapiEngine({
      engine: req.engine,
      defaultEngine,
      searchType: req.searchType,
      extra,
    });
    const params: Record<string, string | number | boolean | undefined> = {
      engine,
      q: req.query,
      num: req.maxResults,
      api_key: ctx.apiKey,
    };
    applySerpapiLocale(params, engine, req.country, req.language);
    applySerpapiSafeSearch(params, engine, req.safeSearch);
    for (const [k, v] of Object.entries(extra)) params[k] = v as string;
    return { method: "GET", url: `${ctx.baseUrl}/search`, headers: { Accept: "application/json" }, params };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = (res.data ?? {}) as Record<string, unknown>;
    const items = extractSerpapiResults(data, req.searchType);
    const results = items.map((item: any) =>
      result({
        title: item.title,
        url: item.link ?? item.url,
        snippet: item.snippet ?? item.description,
        source: item.source ?? item.displayed_link,
        publishedDate: item.date,
        score: item.position,
        raw: item,
      }),
    );
    const answer = extractSerpapiAnswer(data);
    const total = (data.search_information as { total_results?: number } | undefined)?.total_results;
    return finalize(serpapi, req, results, data, { answer, totalResults: total, elapsedMs });
  },
};

export { SERPAPI_WEB_ENGINES };
